### PR TITLE
Fix CAS redirect url

### LINKED ISSUE #2639: CAS validation issue with synapse
Hello,

### Description

It seems there is a problem with how CAS validation is handled in [login.py#L420](https://github.com/matrix-org/synapse/blob/632baf799ea876d2346e934dce90d6a24cb92e37/synapse/rest/client/v1/login.py#L420).

### Steps to reproduce

- Install and configure CAS 5.1.x (I didn't try with earlier ones)
- Install matrix-synapse
- Configure matrix-synapse to use CAS to connect with `service_url: https://matrix.domain`
- Try to login!

You'll see it doesn't work and if I'm not mistaken it is because in [login.py#L397](https://github.com/matrix-org/synapse/blob/632baf799ea876d2346e934dce90d6a24cb92e37/synapse/rest/client/v1/login.py#L397) we ask for a ticket with a service in the form of `https://matrix.domain/_matrix/client/api/v1/login/cas/ticket?redirectUrl=***`.
But when we validate the ticket, we only send `cas_service_url` which lacks the `/_matrix/client/api/v1/login/cas/ticket?redirectUrl=***` part so CAS fails the validation saying the ticket was used for another service than the one it was created for.

I don't know if earlier versions were more lenient on validation, but I got it to work by replacing line 420 by this line:
```python
"service": self.cas_service_url + request.uri[:request.uri.find("&ticket")]
```

But I'm sure there's a better way to do it...

I believe #2404 may be related to this problem.

### Version information

- **Homeserver**: local homeserver

If not matrix.org:
- **Version**: matrix-synapse 0.24.1-1    
- **Install method**: Debian repository
- **Platform**: VM on Debian9

### PR DESCRIPTION
**Pull request by Naugrimm**
Build the same service URL when requesting the CAS ticket and when calling the proxyValidate URL.

<strike>This now of course breaks the CAS-related tests.</strike>
The fixed tests are in https://github.com/matrix-org/sytest/pull/829

Fixes #2639 

Signed-off-by: Erik <erik.gto@googlemail.com>

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

File: synapse/rest/client/v1/login.py [hunk 1]
```diff
@@ -448,7 +448,10 @@ class CasTicketServlet(RestServlet):
     def __init__(self, hs):
         super(CasTicketServlet, self).__init__()
         self.cas_server_url = hs.config.cas_server_url
-        self.cas_service_url = hs.config.cas_service_url
+        self.cas_service_url = (
```
Comment (richvdh) [inline synapse/rest/client/v1/login.py]: it's confusing that `cas_service_url` means something different here than it does in `CasRedirectServlet`.

File: changelog.d/6634.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+Build the same service URL when requesting the CAS ticket and when calling the proxyValidate URL.
```
Comment (richvdh) [inline changelog.d/6634.bugfix]: Could you rephrase this in a way that will make more sense to users? Also, feel free to add an attribution here:

```suggestion
Fix single-sign on with CAS systems: pass the same service URL when requesting the CAS ticket and when calling the `proxyValidate` URL. Contributed by @Naugrimm.

```

maybe?

**Comment by Naugrimm** [comment]
@richvdh I uploaded the changes you requested here as well as the tests in the sytest repo.

File: synapse/rest/client/v1/login.py [hunk 3]
```diff
@@ -73,6 +73,14 @@ def login_id_thirdparty_from_phone(identifier):
     return {"type": "m.id.thirdparty", "medium": "msisdn", "address": msisdn}
 
 
+def build_service_param(cas_service_url, client_redirect_url):
+    return "%s%s?redirectUrl=%s" % (
+        cas_service_url,
+        "/_matrix/client/r0/login/cas/ticket",
+        urllib.parse.quote(client_redirect_url, safe=""),
```
Comment (richvdh) [inline synapse/rest/client/v1/login.py]: ideally we'd use `urllib.parse.urlencode` here; I think this is near enough

File: synapse/rest/client/v1/login.py [hunk 4]
```diff
@@ -428,18 +436,15 @@ def get_sso_url(self, client_redirect_url):
...
-        hs_redirect_url = self.cas_service_url + b"/_matrix/client/r0/login/cas/ticket"
-        service_param = urllib.parse.urlencode(
-            {b"service": b"%s?%s" % (hs_redirect_url, client_redirect_url_param)}
-        ).encode("ascii")
-        return b"%s/login?%s" % (self.cas_server_url, service_param)
+        args = urllib.parse.urlencode(
+            {"service": build_service_param(self.cas_service_url, client_redirect_url)}
+        )
+
+        return "%s/login?%s" % (self.cas_server_url, args)
```
Comment (richvdh) [inline synapse/rest/client/v1/login.py]: according to `BaseSSORedirectServlet`, `get_sso_url` is supposed to return a `bytes`, not a `str`. I think it works because `Request.redirect` works with a `str`, but ideally we'd follow the documentation.
